Bluetooth: ISO: Add check for pending ISO connections

Thalley opened the pull request:
Per the HCI spec it is not possible to send the
HCI_LE_Create_CIS command while an outstanding request
is pending. To avoid failing the command we can
check and verify if any ISO connections are in the
pending state.
    
Since multiple ISO channels can be connected
in a single request, the solution implemented was
to iterate over all ISO connections and see if there was
any unicast ISO channels in the connecting state.
    
It also checks for the encryption pending state, as
we should not start the encryption procedure if we know
that the connect ISO command cannot be completed afterwards.

This adds a new return value to bt_iso_chan_connect,
and decided to properly document the return values the
function can return.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/46829

Thalley committed:
Bluetooth: ISO: Add check for pending ISO connections

Per the HCI spec it is not possible to send the
HCI_LE_Create_CIS command while an outstanding request
is pending. To avoid failing the command we can
check and verify if any ISO connections are in the
pending state.

Since multiple ISO channels can be connected
in a single request, the solution implemented was
to iterate over all ISO connections and see if there was
any unicast ISO channels in the connecting state.

It also checks for the encryption pending state, as
we should not start the encryption procedure if we know
that the connect ISO command cannot be completed afterwards.

This adds a new return value to bt_iso_chan_connect,
and decided to properly document the return values the
function can return.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/iso.h:
```diff
@@ -618,7 +618,26 @@ int bt_iso_cig_terminate(struct bt_iso_cig *cig);
  *  @param count Number of connect parameters.
  *
- *  @return 0 in case of success or negative value in case of error.
+ *  @retval 0 Successfully started the connecting procedure.
+ *
+ *  @retval -EINVAL Invalid parameters were supplied.
+ *
+ *  @retval -EBUSY Some ISO channels are already being connected.
+ *          It is not possible to have multiple outstanding connection requests.
+ *          May also be returned if @kconfig{CONFIG_BT_SMP} is enabled and a
+ *          pairing procedure is already in progress.
+ *
+ *  @retval -ENOBUFS Not buffers available to send request to controller or if
+ *          @kconfig{CONFIG_BT_SMP} is enabled and no more keys could be stored.
+ *
+ *  @retval -ENOMEM If @kconfig{CONFIG_BT_SMP} is enabled and no more keys
+ *          could be stored.
+ *
+ *  @retval -EIO Controller rejected the request or if @kconfig{CONFIG_BT_SMP}
+ *          is enabled and pairing has timed out.
+ *
+ *  @retval -ENOTCONN If @kconfig{CONFIG_BT_SMP} is enabled the ACL is not
+ *          connected.
```
This could be simplied if we have an easy way of saying "or the return values from `bt_conn_set_security`. 

If anyone can suggest a wording where we can rely on the documentation from other (public) functions (such as `bt_conn_set_security`) that would be greatly appreciated (and would also make it easier to keep up to date if the return values of `bt_conn_set_security` changes).

alwa-nordic commented:
I don't have the necessary understanding of ISO to review this PR in its current form. If you would like to me to, you will have to make the description more accessible.

Thalley commented:
> Than you for the explanation. But I still don't understand what the purpose of this change is. AFAIU, without this change, the controller will just deny the HCI command. Isn't that ok? 

Technically it is OK, but we would much rather avoid sending the HCI command if we know it will fail. We do that for almost all HCI commands in the host (and this is the primary issue that is being solved). 

> You are skirting around the "(or encrypting)"-bit like it's not important, but I suspect that this is the real reason behind this change. 

To some extend, it is true that this is one of the reasons behind this change, i.e. there's no reason to start encryption if we know the next procedure (i.e. actually sending the HCI command) will fail, but this is not the main reason. 

> The commit message / PR description needs some love. Please rewrite it so it's obvious what motivates this change.

The motivation of this change is that we don't want to send an HCI command that we know will fail. I think this is already covered by the commit message by

> Per the HCI spec it is not possible to send the HCI_LE_Create_CIS command while an outstanding request is pending. Since multiple ISO channels can be connected in a single request, the solution implemented was to iterate over all ISO connections and see if there was any unicast ISO channels in the connecting state.

 
> > Per the HCI spec it is not possible to send the HCI_LE_Create_CIS command while an outstanding request is pending.
> 
> I found this sentence in the spec: "If the Host issues this command before all the HCI_LE_CIS_Established events from the previous use of the command have been generated, the Controller shall return the error code Command Disallowed (0x0C)." Is that all you are referring to?

Yes

> > Since multiple ISO channels can be connected in a single request, the solution implemented was to iterate over all ISO connections and see if there was any unicast ISO channels in the connecting state.
> 
> Non sequitur. What was the problem that needed a solution? The fact that 'multiple ISO channels can be connected in a single request' does not describe a problem, nor does it motivate the solution.
 
That part describes the solution, and not the problem. The problem is stated earlier in the commit message. 

> > It also checks for the encryption pending state, which is an additional check to make it simpler to handle in the ISO layer, as once security have changed we would not want to have to check for other pending ISO connections.
> 
> The construction of this sentence is really awkward. It should be split up and reworded. What is the ISO layer? The whole change is in `iso.c`, so I assume you have something more specific in mind. Please explain the problem carefully and why the solution works. What security change? (I actually don't know.) Why do we need to care if there are other "pending" ISO connections when security changes? How messy would it be if we did the check? Why do we not need the check with the change in PR?

I can rewrite this paragraph, and add additional information - I do, however, not want to explain how the entire ISO architecture is designed and how it works as part of a commit that doesn't change any behavior except returning an error earlier by not sending a command that is known to fail to the controller. 

In this context `iso.c` is the ISO layer, and I used that term because ISO is implemented as a layer on top of `conn.c` (sort of as an extension of conn.c for a specific type of connection), but it may have been the wrong term to use :)

Thalley commented:
@alwa-nordic Maybe it will be simpler to understand, and be a simpler implementation, if we just added a another `BT_DEV_` flag (e.g. `BT_DEV_ISO_INITIATING`), and simplified the check to be

```c
if (atomic_test_bit(bt_dev.flags, BT_DEV_ISO_INITIATING)) {
	return -EALREADY;
}
```

So that it will have the exact same behavior as the check we have for ACL?

Thalley commented:
I also updated the commit message slightly
